source/server/connection_handler_impl.cc

```diff
@@ -185,30 +185,43 @@ void ConnectionHandlerImpl::ActiveListener::onAccept(
 }
 
 void ConnectionHandlerImpl::ActiveListener::newConnection(Network::ConnectionSocketPtr&& socket) {
-  Network::ConnectionPtr new_connection = parent_.dispatcher_.createServerConnection(
-      std::move(socket), config_.transportSocketFactory().createTransportSocket());
+  // Find matching filter chain.
+  const auto filter_chain = config_.filterChainManager().findFilterChain(
+      const_cast<Network::ConnectionSocket*>(socket.get()));
+  if (filter_chain == nullptr) {
+    ENVOY_LOG_TO_LOGGER(parent_.logger_, debug,
+                        "closing connection: no matching filter chain found");
+    stats_.no_filter_chain_match_.inc();
+    socket->close();
+    return;
+  }
+
+  auto transport_socket = filter_chain->createTransportSocket();
+  Network::ConnectionPtr new_connection =
+      parent_.dispatcher_.createServerConnection(std::move(socket), std::move(transport_socket));
   new_connection->setBufferLimits(config_.perConnectionBufferLimitBytes());
+
+  const bool empty_filter_chain = !config_.filterChainFactory().createNetworkFilterChain(
+      *new_connection, filter_chain->getNetworkFilterFactories());
+  if (empty_filter_chain) {
+    ENVOY_CONN_LOG_TO_LOGGER(parent_.logger_, debug, "closing connection: no filters",
+                             *new_connection);
+    new_connection->close(Network::ConnectionCloseType::NoFlush);
```

Comment:
Also bump stat on this one?

Reply:
I intentionally didn't add it here, since this can only happen if filter chain is intentionally configured without filters (i.e. blackhole), therefore I don't see a value in stat here.

But feel free to disagree, and I'll add one.